File: src/orchestrator/queue-bridge.ts
import { existsSync, readFileSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import type { QueuedTask } from '../queue.js';
import { listTasks, setTaskStatus } from '../queue.js';
import type { ProjectBoard, OrchestratorTask } from './types.js';
import { loadBoard, saveBoard, createBoard } from './store.js';
import { fallbackPlan, parsePlan } from './planner.js';
import type { WorkerName } from '../types.js';
/**
 * Bridge: turn a queued idea (goal string) into a board DAG via the planner,
 * or flatten board tasks back to queue items for Orca workers watching the queue.
 * Mirrors store.ts atomic writes, never logs secrets.
 *
 * Cross-board retry memory (Q27): when a stuck board is archived (orchestrate-loop
 * moves it to .devagent/archive/), its executor failure class (task.interrupt.failureClass,
 * PRD:775 / Q24 taxonomy) is carried onto the re-bridged board so the scout
 * deprioritizes the goal until the root cause ships instead of burning a fresh
 * attempt budget on each requeue round.
 */

/** Where stuck/completed boards are archived by the orchestrate loop. */
const ARCHIVE_DIR = '.devagent/archive';

/** Normalize goal text for cross-board matching (mirror already_shipped in selfbuild-loop.sh). */
function normalizeGoal(goal: string): string {
  return goal.replace(/"/g, '').replace(/\s+/g, ' ').trim();
}

/**
 * Prior board's executor failure class for a re-bridged goal. Scans archived
 * boards (newest first) for one whose goal matches `goal`; returns its carried
 * failureClass, else the first task's interrupt failure class. Best-effort:
 * missing/corrupt archive entries are skipped.
 */
export function archivedBoardFailureClass(repoPath: string, goal: string): string | undefined {
  const archiveDir = join(repoPath, ARCHIVE_DIR);
  if (!existsSync(archiveDir)) return undefined;
  let files: string[];
  try {
    files = readdirSync(archiveDir).filter((f) => f.startsWith('board-') && f.endsWith('.json')).sort().reverse();
  } catch {
    return undefined;
  }
  const want = normalizeGoal(goal);
  for (const file of files) {
    let board: ProjectBoard;
    try {
      board = JSON.parse(readFileSync(join(archiveDir, file), 'utf8')) as ProjectBoard;
    } catch {
      continue; // corrupt archive entry — skip, never fail the bridge
    }
    if (!board || typeof board.goal !== 'string') continue;
    const have = normalizeGoal(board.goal);
    if (want !== have && want.slice(0, 60) !== have.slice(0, 60)) continue;
    if (board.failureClass) return board.failureClass;
    const interrupted = board.tasks?.find((t) => t.interrupt?.failureClass);
    if (interrupted?.interrupt?.failureClass) return interrupted.interrupt.failureClass;
  }
  return undefined;
}

export interface BridgeOptions {
  repoPath: string;
  /** Boards are keyed by the queued goal so re-runs are idempotent. */
  boardGoalKey?: (task: QueuedTask) => string;
  /** How the planner is invoked for a goal; tests inject a stub. */
  planner?: (goal: string) => Promise<OrchestratorTask[] | null>;
}

function defaultPlanner(goal: string): Promise<OrchestratorTask[] | null> {
  // Synchronous fallback-parse never needs a worker: wrap deterministic parse + fallback
  return Promise.resolve(fallbackPlan(goal));
}

export function queuedGoalString(t: QueuedTask): string {
  const ac = t.acceptanceCriteria?.length ? `\nAcceptance criteria:\n${t.acceptanceCriteria.map((c) => `- ${c}`).join('\n')}` : '';
  return `${t.goal}${ac}`;
}

export interface BridgeResult {
  boardPath: string;
  tasksWritten: number;
  created: boolean;
  /** Tasks were already present for this goal-backed board (idempotency) */
  idempotent: boolean;
}

/** Idempotent: if a board already exists for this repo, reuse it; else plan one queued goal into a board. */
export async function bridgeQueueToBoard(
  oldestPending: QueuedTask,
  opts: BridgeOptions,
): Promise<BridgeResult> {
  const repoPath = opts.repoPath;
  const boardPath = join(repoPath, '.devagent-project.json');
  const existing = loadBoard(repoPath);
  if (existing) {
    return { boardPath, tasksWritten: existing.tasks.length, created: false, idempotent: true };
  }
  const goal = queuedGoalString(oldestPending);
  const planner = opts.planner ?? defaultPlanner;
  const planned = await planner(goal);
  let tasks: OrchestratorTask[] = planned && Array.isArray(planned) && planned.length > 0 ? planned : fallbackPlan(goal);
  if (tasks.length > 12) tasks = tasks.slice(0, 12);

  const roles: ProjectBoard['roles'] = { planner: 'omp' as WorkerName, executor: 'omp' as WorkerName };
  const board = createBoard(goal, tasks, roles);
  // Q27 cross-board retry memory: a re-bridged goal carries the prior archived
  // board's executor failure class so the scout deprioritizes it until the root
  // cause ships, instead of burning a fresh attempt budget each requeue round.
  const priorFailureClass = archivedBoardFailureClass(repoPath, goal);
  if (priorFailureClass) board.failureClass = priorFailureClass;
  saveBoard(repoPath, board);
  // Retire the source queue item: the board now owns this goal, and leaving it
  // pending would make the builder lane double-build the same idea.
  setTaskStatus(repoPath, oldestPending.id, 'done');
  return { boardPath, tasksWritten: tasks.length, created: true, idempotent: false };
}

/** Sync dry path over the live queue: pick oldest pending and bridge it (no-op when board exists). */
export async function bridgeIfQueued(repoPath: string, planner?: BridgeOptions['planner']): Promise<BridgeResult | null> {
  const pending = listTasks(repoPath).filter((t: QueuedTask) => t.status === 'pending').sort((a: QueuedTask, b: QueuedTask) => a.createdAt.localeCompare(b.createdAt));
  if (pending.length === 0) return null;
  return bridgeQueueToBoard(pending[0]!, { repoPath, planner });
}

/** Evidence source check for tests: does progress.md currently want a board section? */
export function hasBoard(repoPath: string): boolean {
  return existsSync(join(repoPath, '.devagent-project.json'));
}

File: internal/orchestrator/queue-bridge.go
// Package file mirrors src/orchestrator/queue-bridge.ts (FR-GO-07, issue #194).
package orchestrator

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/FreePeak/devagent/internal/queue"
)

// BridgeArchiveDir mirrors the TS ARCHIVE_DIR: where stuck/completed boards
// are archived by the orchestrate loop.
const BridgeArchiveDir = ".devagent/archive"

// normalizeGoal mirrors the TS normalizeGoal: normalize goal text for
// cross-board matching (mirror already_shipped in selfbuild-loop.sh).
func normalizeGoal(goal string) string {
	return strings.TrimSpace(strings.Join(strings.Fields(strings.ReplaceAll(goal, `"`, "")), " "))
}

// ArchivedBoardFailureClass mirrors archivedBoardFailureClass: a prior
// board's executor failure class for a re-bridged goal. Scans archived
// boards (newest first) for one whose goal matches `goal`; returns its
// carried failureClass, else the first task's interrupt failure class.
// Best-effort: missing/corrupt archive entries are skipped. Empty string is
// the TS undefined.
func ArchivedBoardFailureClass(repoPath, goal string) string {
	archiveDir := filepath.Join(repoPath, BridgeArchiveDir)
	if _, err := os.Stat(archiveDir); err != nil {
		return ""
	}
	entries, err := os.ReadDir(archiveDir)
	if err != nil {
		return ""
	}
	var files []string
	for _, e := range entries {
		name := e.Name()
		if strings.HasPrefix(name, "board-") && strings.HasSuffix(name, ".json") {
			files = append(files, name)
		}
	}
	sort.Strings(files) // then iterate newest first
	want := normalizeGoal(goal)
	for i := len(files) - 1; i >= 0; i-- {
		var board struct {
			Goal         *string            `json:"goal"`
			FailureClass string             `json:"failureClass"`
			Tasks        []OrchestratorTask `json:"tasks"`
		}
		data, err := os.ReadFile(filepath.Join(archiveDir, files[i]))
		if err != nil || json.Unmarshal(data, &board) != nil {
			continue // corrupt archive entry — skip, never fail the bridge
		}
		if board.Goal == nil {
			continue
		}
		have := normalizeGoal(*board.Goal)
		if want != have && firstChars(want, 60) != firstChars(have, 60) {
			continue
		}
		if board.FailureClass != "" {
			return board.FailureClass
		}
		for _, t := range board.Tasks {
			if t.Interrupt != nil && t.Interrupt.FailureClass != "" {
				return t.Interrupt.FailureClass
			}
		}
	}
	return ""
}

// BridgeOptions mirrors the TS BridgeOptions.
type BridgeOptions struct {
	RepoPath string
	// BoardGoalKey: boards are keyed by the queued goal so re-runs are
	// idempotent. (Declared in the TS too; the bridge derives the key from
	// the queued goal itself.)
	BoardGoalKey func(task *queue.QueuedTask) string
	// Planner: how the planner is invoked for a goal; tests inject a stub.
	// Nil = the deterministic fallback plan.
	Planner func(goal string) []OrchestratorTask
}

// defaultPlanner mirrors the TS defaultPlanner: the synchronous
// fallback-parse never needs a worker.
//
// TODO(FR-GO-07): delegate to the planner.go port of fallbackPlan when it
// lands (planner.ts is outside this sub-wave's ownership list); the body
// below is a verbatim copy of TS fallbackPlan.
func defaultPlanner(goal string) []OrchestratorTask {
	return []OrchestratorTask{{
		ID:        "T1",
		Title:     fmt.Sprintf("Implement goal: %s", firstChars(goal, 80)),
		Prompt:    goal,
		DependsOn: []string{},
		Status:    TaskStatusPending,
		Attempts:  0,
	}}
}

// QueuedGoalString mirrors queuedGoalString: the goal a queued task bridges
// with, acceptance criteria appended as a dash list.
func QueuedGoalString(t *queue.QueuedTask) string {
	ac := ""
	if len(t.AcceptanceCriteria) > 0 {
		parts := make([]string, len(t.AcceptanceCriteria))
		for i, c := range t.AcceptanceCriteria {
			parts[i] = "- " + c
		}
		ac = "\nAcceptance criteria:\n" + strings.Join(parts, "\n")
	}
	return t.Goal + ac
}

// BridgeResult mirrors the TS BridgeResult.
type BridgeResult struct {
	BoardPath    string `json:"boardPath"`
	TasksWritten int    `json:"tasksWritten"`
	Created      bool   `json:"created"`
	// Idempotent: tasks were already present for this goal-backed board.
	Idempotent bool `json:"idempotent"`
}

// BridgeQueueToBoard mirrors bridgeQueueToBoard: idempotent — if a board
// already exists for this repo, reuse it; else plan one queued goal into a
// board. Q27 cross-board retry memory: a re-bridged goal carries the prior
// archived board's executor failure class so the scout deprioritizes it
// until the root cause ships. Retires the source queue item: the board now
// owns this goal, and leaving it pending would make the builder lane
// double-build the same idea.
func BridgeQueueToBoard(oldestPending *queue.QueuedTask, opts BridgeOptions) BridgeResult {
	repoPath := opts.RepoPath
	boardPath := filepath.Join(repoPath, BoardFile)
	if existing := LoadBoard(repoPath); existing != nil {
		return BridgeResult{BoardPath: boardPath, TasksWritten: len(existing.Tasks), Created: false, Idempotent: true}
	}
	goal := QueuedGoalString(oldestPending)
	planner := opts.Planner
	if planner == nil {
		planner = defaultPlanner
	}
	planned := planner(goal)
	tasks := planned
	if len(tasks) == 0 {
		tasks = defaultPlanner(goal) // TS: fallbackPlan(goal)
	}
	if len(tasks) > 12 {
		tasks = tasks[:12]
	}
	roles := ProjectBoardRoles{Planner: "omp", Executor: "omp"}
	board := CreateBoard(goal, tasks, roles)
	if prior := ArchivedBoardFailureClass(repoPath, goal); prior != "" {
		board.FailureClass = prior
	}
	SaveBoard(repoPath, board)
	// Unfenced write, mirroring the TS setTaskStatus default.
	queue.SetTaskStatus(repoPath, oldestPending.ID, queue.StatusDone, "", -1, nil)
	return BridgeResult{BoardPath: boardPath, TasksWritten: len(tasks), Created: true, Idempotent: false}
}

// BridgeIfQueued mirrors bridgeIfQueued: sync dry path over the live queue —
// pick oldest pending and bridge it (no-op when board exists). A nil
// planner means no bridge happens when the queue is empty either way.
func BridgeIfQueued(repoPath string, planner func(goal string) []OrchestratorTask) *BridgeResult {
	pending := queue.ListTasks(repoPath, queue.StatusPending) // sorted by createdAt ascending
	if len(pending) == 0 {
		return nil
	}
	res := BridgeQueueToBoard(pending[0], BridgeOptions{RepoPath: repoPath, Planner: planner})
	return &res
}

// HasBoard mirrors hasBoard: does .devagent-project.json currently exist?
func HasBoard(repoPath string) bool {
	_, err := os.Stat(filepath.Join(repoPath, BoardFile))
	return err == nil
}
